wallet: Replace use of purpose strings with an enum

achow101 opened the pull request:
Instead of storing and passing around fixed strings for the purpose of an address, use an enum.

achow101 committed:
wallet: Replace use of purpose strings with an enum

Instead of storing and passing around fixed strings for the purpose of
an address, use an enum.

This also rationalizes the CAddressBookData struct, documenting all fields and
making them public, and simplifying the representation to avoid bugs like
https://github.com/bitcoin/bitcoin/pull/26761#discussion_r1134615114 and make
it not possible to invalid address data like change addresses with labels.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

maflcko reviewed (commented):
nits

achow101 commented on test/functional/wallet_labels.py:
```diff
@@ -71,6 +71,10 @@ def run_test(self):
 
+        self.log.info("Checking listlabels' invalid parameters")
+        assert_raises_rpc_error(-8, "Invalid 'purpose' argument, must be one of 'send', or 'receive'.", node.listlabels, "notavalidpurpose");
+        assert_raises_rpc_error(-8, "Invalid 'purpose' argument, must be one of 'send', or 'receive'.", node.listlabels, "unknown");
```
Fixed

achow101 commented on src/wallet/wallet.h:
```diff
@@ -206,10 +207,48 @@ class CAddressBookData
+        case interfaces::Wallet::AddressPurpose::SEND:
+            return "send";
+        // no default case so the compiler will warn when a new enum as added
+        }
```
Done

achow101 commented on src/wallet/wallet.h:
```diff
@@ -206,10 +207,48 @@ class CAddressBookData
+        } else if (s == "send") {
+            return interfaces::Wallet::AddressPurpose::SEND;
+        } else {
+            // Any other string is a programming error
```
Done

ryanofsky reviewed (commented):
This is a nice change, and I started reviewing 44a2ae100b600bc9c9f4c956223fab3b1948ea8d, but it looks like there are a lot of functions taking `optional<AddressPurpose>` values that should more properly take `AddressPurpose` values for simplicity and better type checking. (Basically `NotifyAddressBookChanged` and all the functions it calls). I think this might stem from not having separate SetAddressBook and UpdateAddressBook functions, so I want to look into this before reviewing more of the PR. (Having separate functions could help because Set function could make purpose a required parameter, while the Update function could let it be it optional).

ryanofsky committed:
wallet: Add wallet/types.h for simple public enum and struct types

Move isminetype and isminefilter there this commit, add WalletPurpose type next
commit.

ryanofsky committed:
wallet: add AddressPurpose enum to replace string values

ryanofsky commented:
I made some changes to this in a branch that could be adopted https://github.com/ryanofsky/bitcoin/commits/review.27217.1-edit

- Tries to split commits and add a lot of comments so this is easier to review
- Make enum a public wallet type instead of putting it in interfaces/ so wallet/ code does not depend on interfaces/wallet code and `interfaces::Wallet::AddressPurpose` becomes just `wallet::AddressPurpose`
- Replaced `std::optional<AddressPurpose>` with plain `AddressPurpose` in many places where purpose is known or can be derived from IsMine, to simplify code and avoid bugs.
- Dropped `AddressPurpose::UNKNOWN` to avoid defining an enum value that doesn't have any corresponding value in the wallet database.
- Added `AddressPurpse::REFUND` enum value to be able to load old wallets that may have refund addresses without errors or warnings.
- Changed assert(0) failure when loading wallets with unrecognized purpose strings with a warning.

achow101 commented:
> I made some changes to this in a branch that could be adopted https://github.com/ryanofsky/bitcoin/commits/review.27217.1-edit

I've adopted these changes.

ryanofsky commented on test/functional/wallet_labels.py:
```diff
@@ -71,6 +71,10 @@ def run_test(self):
         node = self.nodes[0]
         assert_equal(len(node.listunspent()), 0)
 
+        self.log.info("Checking listlabels' invalid parameters")
+        assert_raises_rpc_error(-8, "Invalid 'purpose' argument, must be one of 'send', or 'receive'.", node.listlabels, "notavalidpurpose")
+        assert_raises_rpc_error(-8, "Invalid 'purpose' argument, must be one of 'send', or 'receive'.", node.listlabels, "unknown")
```
In commit "wallet: Replace use of purpose strings with an enum" (db521d3e89b23046f73410a875bb9660178e3842)

I think the PR might benefit from a release note, or at least a summary of behavior changes in the PR description like, 'Wallet loading code will now issue warnings if addresses have unrecognized purpose strings other than "send" "receive" and "refund" and `listlabels` RPC will raise an error if an unrecognized purpose is requested.'

achow101 replied:
Done

ryanofsky reviewed (approved):
Code review ACK db521d3e89b23046f73410a875bb9660178e3842. Thanks for adopting suggested changes. I left some new suggestions, but they are not important and can be ignored.

I think this PR should be straightforward for others to review now, and nice because it should demystify the purpose field and wallet address book as a whole.

achow101 committed:
doc: Release note for purpose string restriction

achow101 commented on src/wallet/interfaces.cpp:
```diff
@@ -189,29 +189,34 @@ class WalletImpl : public Wallet
         }
         if (purpose) {
-            *purpose = entry->purpose;
+            *purpose = entry->purpose ? *entry->purpose : *dest_is_mine ? AddressPurpose::RECEIVE : AddressPurpose::SEND;
```
Done

achow101 commented on src/wallet/interfaces.cpp:
```diff
@@ -189,29 +189,34 @@ class WalletImpl : public Wallet
             if (is_change) return;
-            result.emplace_back(dest, m_wallet->IsMine(dest), label, purpose);
+            isminetype is_mine = m_wallet->IsMine(dest);
+            result.emplace_back(dest, is_mine, purpose ? *purpose : is_mine ? AddressPurpose::RECEIVE : AddressPurpose::SEND, label);
```
Done

achow101 commented on src/wallet/rpc/addresses.cpp:
```diff
@@ -721,9 +721,15 @@ RPCHelpMan listlabels()
-        purpose = request.params[0].get_str();
+        std::string purpose_str = request.params[0].get_str();
+        if (!purpose_str.empty()) {
+            if (!IsValidPurposeString(purpose_str)) {
```
Done

achow101 commented on src/wallet/types.h:
```diff
@@ -57,7 +57,7 @@ using isminefilter = std::underlying_type<isminetype>::type;
  * interfaces and saved for new addresses. It is basically redundant with an
  * address's IsMine() result.
  */
-enum class AddressPurpose : uint8_t {
```
Done

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2399,35 +2399,39 @@ DBErrors CWallet::ZapSelectTx(std::vector<uint256>& vHashIn, std::vector<uint256
     }
     NotifyAddressBookChanged(address, strName, is_mine,
-                             strPurpose, (fUpdated ? CT_UPDATED : CT_NEW));
-    if (!strPurpose.empty() && !batch.WritePurpose(EncodeDestination(address), strPurpose))
+                             purpose ? *purpose : is_mine ? AddressPurpose::RECEIVE : AddressPurpose::SEND,
+                             (fUpdated ? CT_UPDATED : CT_NEW));
```
Done

achow101 commented on src/wallet/wallet.h:
```diff
@@ -200,28 +201,75 @@ class ReserveDestination
+     * Additional address metadata map that can currently hold two types of keys:
+     *
+     *   "used" keys with values always set to "p" if present. This is set on
+     *       IsMine addresses that have already been spend from if the
```
Done

ryanofsky commented on doc/release-notes-27217.md:
```diff
@@ -0,0 +1,6 @@
+Wallet
+------
+
+* Purposes strings are restricted to the currently known values of "send", "receive", and "refund".
```
In commit "doc: Release note for purpose string restriction" (0d18142f595b43070bd07b608d43c9ed9b37cdcc)

Maybe s/restricted/now restricted/ to be clear this is new behavior not background information.

achow101 replied:
Done

ryanofsky reviewed (approved):
Code review ACK 0d18142f595b43070bd07b608d43c9ed9b37cdcc. Suggested a few comment/whitespace tweaks, but this is also fine as-is.

achow101 commented on src/wallet/transaction.h:
```diff
@@ -108,8 +110,30 @@ static inline int TxStateSerializedIndex(const TxState& state)
 typedef std::map<std::string, std::string> mapValue_t;
 
+
+
```
Fixed

ryanofsky reviewed (approved):
Code review ACK a48010f1fd6d8a430b8234789ac3c91ec9d06972. Just comment and whitespace changes since last review

aureleoules reviewed (approved):
ACK a48010f1fd6d8a430b8234789ac3c91ec9d06972, looks good to me.

josibake commented on src/qt/addresstablemodel.cpp:
```diff
@@ -52,16 +53,17 @@ struct AddressTableEntryLessThan
-    if (strPurpose == "send")
+    if (purpose == wallet::AddressPurpose::SEND) {
         addressType = AddressTableEntry::Sending;
-    else if (strPurpose == "receive")
+    } else if (purpose == wallet::AddressPurpose::RECEIVE) {
         addressType = AddressTableEntry::Receiving;
-    else if (strPurpose == "unknown" || strPurpose == "") // if purpose not set, guess
+    } else {
         addressType = (isMine ? AddressTableEntry::Receiving : AddressTableEntry::Sending);
+    }
```
IIRC, don't we get all sorts of compiler goodness if we use `switch` with an enum, like compiler warnings for unhandled cases? Would be nice, in the event we ever add a new case to the enum (although seems unlikely for this specific use-case).

I always prefer using `switch` with enums, but feel free to ignore if you don't feel like retouching/don't share the same opinion.

josibake replied:
https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wswitch-enum , super nice and still warns for unhandled cases even when you have a default.

achow101 replied:
Done as suggested

josibake reviewed (commented):
Code review ACK https://github.com/bitcoin/bitcoin/pull/27217/commits/a48010f1fd6d8a430b8234789ac3c91ec9d06972

Absolutely love this. Death to strings. 

Looks good from a code review standpoint, also did some light testing with the GUI and RPC calls, planning on doing more.

ryanofsky reviewed (commented):
~This PR has 3 acks so maybe it is ready to be merged.~ I think this PR basically looks good as is but it would be a little better if it avoided changing behavior for refund addresses in translateTransactionType (see comment below)

achow101 commented on src/wallet/wallet.h:
```diff
@@ -200,28 +201,70 @@ class ReserveDestination
-    std::string purpose;
+    /**
+     * Address label for sending and receiving addresses, most commonly set to
+     * "", which is the default label. Change addresses do not have labels, and
```
Done as suggested

achow101 commented on src/wallet/wallet.h:
```diff
@@ -200,28 +201,70 @@ class ReserveDestination
+    /**
+     * Additional address metadata map that can currently hold two types of keys:
+     *
+     *   "used" keys with values always set to "p" if present. This is set on
```
Done as suggested

ryanofsky reviewed (approved):
Code review ACK 98821a7aca88ac9401786e848899866b71446460. Just suggested changes since last review, the main one being changing `translateTransactionType` to use a switch statement.

This had 3 reviews before, so it would be nice to have reacks and hopefully merge this soon.

Xekyo commented on src/wallet/wallet.h:
```diff
@@ -200,28 +201,71 @@ class ReserveDestination
-        m_label = label;
-    }
+    /** Accessor methods. */
+    bool IsChange() const { return !label; }
```
I thought here first that if the user does not set a label, we might treat an address as change. However, I realized that an empty string `""` is truthy, while no string being set is falsy, i.e. even if the address label is set to `""` which appears to be the case for any receive addresses, it’ll be truthy.

achow101 replied:
I've made this clearer by using `has_value()`.

Xekyo commented on src/wallet/wallet.h:
```diff
@@ -200,28 +201,71 @@ class ReserveDestination
+
+inline std::optional<AddressPurpose> PurposeFromString(std::string_view s)
+{
+    if (s == "receive") {
+        return AddressPurpose::RECEIVE;
+    } else if (s == "send") {
+        return AddressPurpose::SEND;
+    } else if (s == "refund") {
+        return AddressPurpose::REFUND;
+    }
```
Nit, since each immediately returns, this may be more readable:

```suggestion
    if (s == "receive") return AddressPurpose::RECEIVE;
    if (s == "send") return AddressPurpose::SEND;
    if (s == "refund") return AddressPurpose::REFUND;
```

achow101 replied:
I prefer the `else if` style, but have collapsed this and `PurposeToString` to be more readable.

Xekyo commented on doc/release-notes-27217.md:
```diff
@@ -0,0 +1,6 @@
+Wallet
+------
+
+* Purposes strings are now restricted to the currently known values of "send", "receive", and "refund".
+  Wallets that have unrecognized purpose strings will have loading warnings, and the `listlabels`
+  RPC will raise an error if an unrecognized purpose is requested.
```
Perhaps mention that the “purpose string” appears in the context of addresses, at least for me that would not necessarily have been obvious just from this release note without reviewing the PR.

achow101 replied:
Done.

murchandamus reviewed (approved):
ACK 98821a7aca88ac9401786e848899866b71446460

achow101 commented on src/wallet/types.h:
```diff
@@ -48,6 +48,20 @@ enum isminetype : unsigned int {
+
+/**
+ * Address purpose field that has been been stored with wallet sending and
+ * receiving addreses since BIP70 payment protocol support was added in
```
Fixed.

josibake commented:
ACK https://github.com/bitcoin/bitcoin/pull/27217/commits/635c50a57b3f72f83f1ee681f6a4b7766b8988f6

Xekyo commented:
reACK https://github.com/bitcoin/bitcoin/commit/635c50a57b3f72f83f1ee681f6a4b7766b8988f6

but appears to need rebase

josibake commented:
reACK https://github.com/bitcoin/bitcoin/commit/18fc71a3adee5de0f74ba6ff18f5ed31ba79a646

denavila commented on src/qt/walletmodel.cpp:
```diff
@@ -377,18 +377,17 @@ static void NotifyKeyStoreStatusChanged(WalletModel *walletmodel)
                               Q_ARG(QString, strLabel),
                               Q_ARG(bool, isMine),
-                              Q_ARG(QString, strPurpose),
+                              Q_ARG(wallet::AddressPurpose, purpose),
```
This change appears to trigger this assert when executing "getnewaddress" from the Console:
Assertion failed: (invoked), function NotifyAddressBookChanged, file walletmodel.cpp, line 392.

debug.log shows this error message:
GUI: QMetaMethod::invoke: Unable to handle unregistered datatype 'wallet::AddressPurpose'

fanquake replied:
Thanks, opened an issue here to follow up: https://github.com/bitcoin-core/gui/issues/725.
